### PR TITLE
Migrate FetchProperties values to standard D-Bus properties

### LINKED ISSUE #2816: Drop FetchProperties interface for all kinds of D-Bus objects
* For filesystems, pools, and blockdevs, migrate all properties managed by FetchProperties interface to standard D-Bus properties.
   * General observations:
     * we'll have to designate some of these properties as optional, returning the value as a tuple.These are properties that we can not guarantee that stratisd can "know". We should err on the side of designating properties as optional, so as to avoid finding later that there was an unsuspected way that they could be optional, but now our API can not accommodate that fact. 
     * some of these properties change rapidly. We want to avoid a storm of signals on the D-Bus for, for example, ```FILESYSTEM_USED```. A naive approach is to send out a signal every time we detect a change; we can start with that, but may have to modify that approach if excessive signals become a burden.
  * For filesystems:
    * ```FILESYSTEM_USED``` - optional. ```EmitsChangedSignal::True```. This property may be updated by stratisd whenever a filesystem check occurs. We can build on the existing infrastructure that we're already using for the filesystem Used property to make sure that D-Bus PropertyChanged signals are managed correctly.
   * For blockdevs:
     * ```BLOCKDEV_TOTAL_SIZE_PROP``` - optional or not, I'm not sure. Could we possible construct a pool with a blockdev with an unknown size? I _think_ not. ```EmitsChangedSignal::True```. The total size can change only on a D-Bus method call to extend the blockdev, so handling signals should be straightforward.
   * For pools:
     * ```POOL_ENCRYPTION_KEY_DESC```, ```POOL_CLEVIS_INFO``` - both optional. ```EmitsChangedSignal:True```. I believe that neither can be updated except by an API call, so handling PropertyChanged signals should be straightforward.
     * ```POOL_HAS_CACHE_PROP``` - we can probably define this as not optional safely. ```EmitsChangedSignal:True```. Updated by API call, so handling PropertyChanged signals should be straightforward.
     * ```POOL_TOTAL_SIZE_PROP``` - this is the sum of the total size of all the data devices in the pool. If we consider ```BLOCKDEV_TOTAL_SIZE_PROP``` optional, then this should also be optional, otherwise not. ```EmitsChangedSignal::True```. Change can occur only due to API calls like adding block devices or extending existing ones.
     * ```POOL_TOTAL_USED_PROP``` - optional. ```EmitsChangedSignal::True```. This ought to change only on devicemapper events which cause a pool check to be run and pool status to be invoked. The only problem here is that this is an aggregate value and if the pool becomes more complex then more component parts may be added, which may depend on something other than thinpool status information.
* For the top level object the two keys that are supported are ```KEY_LIST_PROP``` and ```LOCKED_POOL_DEVS```. There is an outstanding question whether these two keys should be turned into D-Bus properties or the values returned via D-Bus methods. Right now I'm inclining toward D-Bus methods for both. The reasons are:
  * ```KEY_LIST_PROP``` corresponds to the list of stratis keys in the kernel keyring. We go though stratisd to obtain them, but they are not part of stratisd's state at all. I think that's a good reason to return the value via a dedicated method.
  * ```LOCKED_POOL_DEVS``` is less clear. I'm torn, actually.

By doing this we will have made a commitment: if the size of the GetManagedObjects() result becomes a significant problem, either for stratis-cli or for other users, we will provide additional, more specific D-Bus interfaces to get information that users might need.

### PR DESCRIPTION
**Pull request by jbaublitz**
Closes #2816

### CODE REVIEW

**Comment by jbaublitz** [comment]
@mulkieran All tests passed other than the expected failure. This is ready for review.

**Comment by mvollmer** [comment]
Thanks, I got this to work quite nicely with Cockpit here: https://github.com/cockpit-project/cockpit/pull/16532

The "key description" now seems to be doubly optional.  It has a signature of  "(b(bs))" as a pool property, and is encoded like this in the LockedPools property:

```
{ "t": "(bv)", "v": [true, {"t": "(bs)", "v": [true,"pool0"]}]}
```

Can you explain what this means?  What is the difference between `[ false, [ false, "" ]` and `[ true, [ false, "" ]`?

**Comment by jbaublitz** [comment]
> Thanks, I got this to work quite nicely with Cockpit here: [cockpit-project/cockpit#16532](https://github.com/cockpit-project/cockpit/pull/16532)
> 
> The "key description" now seems to be doubly optional. It has a signature of "(b(bs))" as a pool property, and is encoded like this in the LockedPools property:
> 
> ```
> { "t": "(bv)", "v": [true, {"t": "(bs)", "v": [true,"pool0"]}]}
> ```
> 
> Can you explain what this means? What is the difference between `[ false, [ false, "" ]` and `[ true, [ false, "" ]`?

Hi @mvollmer! So the basic idea here is that `StratisResult<Option<KeyDescription>>` is the type internally. If we were to do just a single boolean, it would be unclear whether the key description was set or whether there was an error. We discussed potentially making that a `(bv)` type so that we could make it `(false, [error string])` or `(true, (bool, string))`, but this didn't work very well for two reasons. One is that the formatting that we have in our CLI just doesn't have room for a long error message displaying what went wrong and the other is that Rust still has some error even in the standard library that make cloning/passing around errors impossible. Our wrapper error type inherited this problem as a result. If we were to try to work around this, we would bump into a requirement of stringifying all errors that we're handling in code that needs to report changes to the D-Bus and this gets to be a problem because of the work we've done on establishing an error chain to allow stratisd to react internally to certain errors and drop to a degraded state. Stringifying errors would render this useless. I've done my best to provide this information through other means like our report interface and more robust logging. Just a note that I've already put up a PR for the D-Bus documentation in preparation for 3.0 and I will update it with this information, in particular which booleans mean errors and which mean optional types.

**Comment by mvollmer** [comment]
> If we were to do just a single boolean, it would be unclear whether the key description was set or whether there was an error.

Ah, I see, thanks!

**Comment by mvollmer** [comment]
@jbaublitz, I have tested this now a bit more formally, see https://github.com/cockpit-project/cockpit/pull/16532#issuecomment-958774557. As green as it can be!

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks great! Two comment requests and one note.

File: src/dbus_api/api/manager_3_0/api.rs [hunk 1]
```diff
@@ -44,8 +65,9 @@ where
     f.property::<&str, _>("Version", ())
         .access(Access::Read)
         .emits_changed(EmitsChangedSignal::Const)
-        .on_get(get_version)
+        .on_get(get_version::<E>)
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/dbus_api/api/manager_3_0/api.rs:68]: When I added that, it was necessary. I can remove it if it's no longer necessary.

**Comment by jbaublitz** [comment]
@mulkieran Okay, this should be ready for review!

**Review by mulkieran — APPROVED** [review]

**Comment by jbaublitz** [comment]
@mulkieran Squashed!

**Pull request merged**